### DISCUSSION TITLE
Enhancement of User Progress Tracking and Recent History Features

### DISCUSSION THREAD

**Discussion by abisha-peter** [TheOdinProject/theodinproject#4974] [Ideas]
### Checks

- [X] This is not a duplicate of an existing issue (please have a look through our [open issues list](https://github.com/TheOdinProject/curriculum/issues) to make sure)
- [X] I have thoroughly read and understand [The Odin Project Contributing Guide](https://github.com/TheOdinProject/.github/blob/main/CONTRIBUTING.md)
- [ ] Would you like to work on this issue?

### Describe your suggestion

I would like to propose an enhancement to the user interface of The Odin Project that focuses on improving the tracking of user progress and recent activity. Currently, users may find it challenging to monitor their advancement through the curriculum and to easily revisit recently accessed resources. Implementing a more intuitive progress tracking system that visually represents completed lessons and topics, along with a dedicated section for recent history, could significantly enhance the user experience. This would not only help users stay organized but also motivate them to continue their learning journey by providing a clear overview of their achievements. I believe that by integrating these features, The Odin Project can further empower its users to navigate their educational path more effectively.

### Path

Foundations, Other / NA

### Lesson Url

https://www.theodinproject.com/paths/foundations/courses/foundations

**Comment by JoshDevHub** [comment]
This issue is much more about the site's features than the curriculum's content, so I've transferred it to this repo for further discussion.

**Comment by abisha-peter** [comment]
Any reply after discussion

**Comment by KevinMulhern** [comment]
Thanks for the suggestion @abisha-peter. There seems to be two different pitches within this:

1. "A more intuitive progress tracking system that visually represents completed lessons and topics"
2. "a dedicated section for recent history"

I think you'll get more traction if those were split into two issues/suggestions so they can be considered individually.

**Comment by JohayerChowdury** [comment]
For 1, it can be similar to leetcode where each lesson have checkmarks or ticks beside them indicating that they have been opened. You can also use a progress circle for each topic to indicate how much of the topic you have finished reading.

**Comment by ChuqiaoHuang** [comment]
Can you please assign this to me?

**Comment by KevinMulhern** [comment]
Thanks for the interest, but theres nothing decided/actionable on this issue yet!

**Comment by mao-sz** [comment]
The first point about progress tracking feels hard to improve IMHO. Lesson completion can already be tracked if you're signed in, and I'm of the opinion that there aren't significant enough benefits of more granular progress tracking to warrant the implementation.

Marking % completion within a lesson based on an assignment checklist would mean deciding how to deal with any changes to assignments, all of which are very challenging and begs the "is it actually solving a real problem?" question:
- What happens if new assignments are added?
  - What happens to progress if a new assignment is placed at the end of the list vs any other position?
- What happens if a "checked" assignment gets substantially changed? What constitutes a substantial change? Fixing a typo? Changing the link href? What if a step contains more than one link?
- This would likely also require storing individual assignment steps to the db (if checks are actually assigned to a particular assignment), which seems a real big change for questionable benefit.
  
Scroll-based progress also raises concerns:
- What happens when someone scrolls down far e.g. to the bottom then scrolls back up to where they actually were? Should it be marked as complete? Or should progress reverse? If it reverses, what if someone completes a lesson then later revisits a particular section then exits? How does progress differentiate?
